piker/_daemon.py

```diff
@@ -0,0 +1,141 @@
+"""
+pikerd daemon lifecylcle & rpc
+"""
+from typing import Optional
+from contextlib import asynccontextmanager
+
+import tractor
+
+from .log import get_logger, get_console_log
+from .brokers import get_brokermod
+
+
+log = get_logger(__name__)
+
+_root_nursery: Optional[tractor._trionics.ActorNursery] = None
+_root_dname = 'pikerd'
+_root_modules = [
+    __name__,
+    'piker.clearing._ems',
+    'piker.clearing._client',
+]
+
+
+@asynccontextmanager
+async def open_pikerd(
+    loglevel: Optional[str] = None,
+    **kwargs,
+) -> Optional[tractor._portal.Portal]:
+
+    global _root_nursery
+
+    # XXX: this may open a root actor as well
+    async with tractor.open_nursery(
+        name=_root_dname,
+
+        # TODO: eventually we should be able to avoid
+        # having the root have more then permissions to
+        # spawn other specialized daemons I think?
+        # enable_modules=[__name__],
+        enable_modules=_root_modules,
+
+        loglevel=loglevel,
+    ) as nursery:
+        _root_nursery = nursery
+        yield nursery
+
+
+@asynccontextmanager
+async def maybe_open_pikerd(
+    loglevel: Optional[str] = None,
+    **kwargs,
+) -> Optional[tractor._portal.Portal]:
+    """If no ``pikerd`` daemon-root-actor can be found,
+    assume that role and return a portal to myself
+
+    """
+    global _root_nursery
+
+    if loglevel:
+        get_console_log(loglevel)
+
+    try:
+        async with tractor.find_actor(_root_dname) as portal:
+            if portal is not None:  # pikerd exists
+                yield portal
+                return
+
+    except RuntimeError:  # tractor runtime not started yet
+        pass
+
+    # assume pikerd role
+    async with open_pikerd(
+        loglevel,
+        **kwargs,
+    ):
+        assert _root_nursery
+
+        # in the case where we're starting up the
+        # tractor-piker runtime stack in **this** process
+        # we want to hand off a nursery for starting (as a sub)
+        # whatever actor is requesting pikerd.
+        yield None
+
+
+# brokerd enabled modules
+_data_mods = [
+    'piker.brokers.core',
+    'piker.brokers.data',
+    'piker.data',
+    'piker.data._buffer'
+]
+
+
+async def spawn_brokerd(
```

Comment:
Do you think it's maybe just worth generalizing this spawn logic now?

For example we can have a `spawn_daemon(name: str)` and then just pass the module list in each case?

Reply:
Yeah that might be worth doing, easier to abstract shared behavoirs

Reply:
Kinda started this with the `_Services` type but we probably want to factor out the `spawn_<service>d()` code somewhat eventually as needed.